docs/content/python/guides/enforcement/field.md

```diff
@@ -0,0 +1,80 @@
+---
+
+{{% callout "Note: 0.20.0 Alpha Feature" %}}
+  This is an API provided by the alpha release of Oso 0.20.0, meaning that it is
+  not yet officially released. You may find other docs that conflict with the
+  guidance here, so proceed at your own risk! If you have any questions, don't
+  hesitate to [reach out to us on Slack](https://join-slack.osohq.com). We're
+  here to help.
+{{% /callout %}}
+
+<div class="pb-10"></div>
+
+# Field-level Enforcement
+
+Perhaps you're creating a user profile page, and in your app, only a user's
+friends can access their last check-in location. To enforce this type of
+authorization, it often makes sense to use **"field-level" enforcement**, by
+explicitly allowing access to certain fields on your domain objects.
+
+Field-level authorization gives you fine-grained control over who can access
+exactly what bits of information. In polar, you can write field-level rules like
+this:
+
+```polar
+allow_field(user, "read", profile, "last_check_in_location") if
+    profile.user in user.friends;
+```
+
+Notice that an `allow_field` rule is just like an `allow` rule, except that it
+takes an additional argument: the field name.
+
+To enforce field-level authorization in your app, you use the `authorize_field`
+method provided by Oso enforcers.
+
+{{% minicallout %}}
+**Note**: You'll need an Oso enforcer to use the `authorize_field` method. Check out
+[how to build one](enforcer.html).
+{{% /minicallout %}}
+
+```python
+def get_last_check_in_location(profile, current_user):
+    oso.authorize_field(current_user, "read", profile, "last_check_in_location");
+    return profile.last_check_in.location
+```
+
+Like `authorize`, `authorize_field` will raise an `AuthorizationError` when the
+user is not allowed to perform the given action. This is an error that you
+should handle globally in your app. You can read more details about this on the
+[Resource-level Enforcement Guide](resource.html#authorization-failure).
+
+Sometimes it is helpful to get
+_all_ fields that a user can access, and for this there is a separate method
+called `authorized_fields`:
+
+```python
+# Serialize only the fields of profile that the current user is allowed to read
+def serialize_profile(profile, current_user):
+    fields = oso.authorized_fields(current_user, "read", profile)
+    return { field: profile[field] for key in fields }
+```
+
+The `authorized_fields` method can be used to send only the fields that the user
+is explicitly allowed to read, or can similarly be used to filter _incoming_
+parameters from a user for a call to, say, an `update` method. In that case, you
+might use an `"update"` action in the call to `authorized_fields`:
+
+```python
+# Filter update_params by the fields on profile that the user can update
+def filter_update_params(profile, raw_update_params, current_user):
+    fields = oso.authorized_fields(current_user, "update", profile)
+    return { field: raw_update_params[field] for key in fields }
+```
```

Comment:
It would be good to include in this guide some advice about handling large number of fields, allowing users to read all fields, etc.

Field level enforcement is a bit unfortunate b/c it really needs to be synced up with the model and feels like it would change much more often than adding new models.

This is more of a product suggestion... but I wonder if we'd consider something like a field group.

Imagine something like:

```
class Profile:
   # Suppose this data could go in the policy too
   FIELD_GROUPS = {
       "anyone": ["name"],
       "friends_only": ["last_check_in_location"],
       "private": ["address"]
    }

# Oso enforcer hooks into field groups somehow, so you call
# authorized_fields in the same way
```

```
# polar

# anyone can read "anyone" fields
allow_field_group(_user, "read", _profile, "anyone");

# only friends can read friends fields
allow_field_group(user, "read", profile, "friends_only") if
    user in profile.friends;

# private fields
allow_field_group(user, "read", profile, "private") if
    user = profile;
```

Then, I can know to update `FIELD_GROUPS` when updating the model & adding a field and I don't need to write rules for every single field in my policy.

Reply:
Could also build the mapping of field to field group in the policy if we feel that's a better place for it.

Reply:
Hmm this is a really good point and has me thinking. I don't think the current API is the final word on field-level enforcement.

For now, we could maybe add an example of an allow rule like this?
```
allow_field(user: User, "read", profile: Profile, field) if
  field in Profile.FRIENDS_ONLY_FIELDS and
  user in profile.friends;
```

Reply:
yeah that's a cool example

Reply:
Added another section to this effect -- might appreciate a read-through in `field.md` if you have time.